packages/gatsby/src/internal-plugins/query-runner/__tests__/pages-writer.js

```diff
@@ -0,0 +1,88 @@
+const _ = require(`lodash`)
+const { writePages, resetLastHash } = require(`../pages-writer`)
+
+const jsonDataPathsFixture = require(`./fixtures/jsonDataPaths.json`)
+const pagesFixture = require(`./fixtures/pages.json`)
+
+jest.mock(`fs-extra`, () => {
+  return {
+    writeFile: (filePath, data) => Promise.resolve(),
+    move: () => {},
+  }
+})
+const mockFsExtra = require(`fs-extra`)
+
+let mockState = {
+  program: {
+    _: [`build`],
+    directory: `my/gatsby/project`,
+  },
+  jsonDataPaths: jsonDataPathsFixture,
+  pages: {
+    values: () => pagesFixture,
+  },
+}
+
+jest.mock(`../../../redux/`, () => {
+  return {
+    store: {
+      getState: () => mockState,
+    },
+    emitter: {
+      on: () => {},
+    },
+  }
+})
+
+describe(`Pages writer`, () => {
+  it(`writes pages with the good order`, async () => {
+    const spy = jest.spyOn(mockFsExtra, `writeFile`)
+
+    await writePages()
+    const data1 = spy.mock.calls[3][1]
```

Comment:
actually - could we make this clearer - in example in `writeFile` function mock store string we are writing if we are writing to `data.json` file, so we don't have to rely on index here (which is hard to read when someone doesn't have context - as in - why exactly are we looking for 3rd and 7th writeFile call)

Reply:
Are you talking about storing `data` object receive in `writeFile` arguments into a global variable and use it?

Something like:

```js
let writeFileCalls = {}

jest.mock(`fs-extra`, () => {
  return {
    writeFile: (filePath, data) => {
      writeFileCalls[filePath] = data;

      return Promise.resolve();
    },
    move: () => {},
  }
})

// ...

await writePages()
const data1 = writeFileCalls['data.json']
```

Reply:
that will work, yeah

Reply:
actually some additional processing would need to be done to path as `writeAndMove` create absolute path - but you got general gist of my request

Reply:
also this is not a blocker - if you feel like don't have time now for it, just let me know and we will merge as-is (just potentially add `// TO-DO` comment pointing here if someone would want to improve it.

Reply:
Actually after implementing the example above, I get an error:

> babel-plugin-jest-hoist: The module factory of `jest.mock()` is not allowed to reference any out-of-scope variables.
> Invalid variable access: writeFileCalls

There certainly is a solution but my knowledges about testing with Jest are too limited.

Reply:
ah yeah, I remember that now - don't worry about it then, we'll merge soonish as-is

Reply:
I found a solution by using `toHaveBeenLastCalledWith` function:

```js
expect(spy).toHaveBeenLastCalledWith(
  `my/gatsby/project/.cache/data.json.${now}`,
  expectedResult
)
```

Reply:
WTY?

Reply:
Can we try `toHaveBeenCalledWith` (without `Last` part) and reset spy between `writePages()` calls?

Reply:
I

Reply:
I have no choice because the call I'm testing is the last of 4 calls. Using `toHaveBeenCalledWith` fails the test.

Reply:
I can use [`toHaveBeenNthCalledWith `](https://jestjs.io/docs/en/expect#tohavebeennthcalledwithnthcall-arg1-arg2-) function otherwise

Reply:
> I have no choice because the call that I'm testing is the last call of 4 successives. Using toHaveBeenCalledWith fails the test.

That's not good :/ Because it can be flaky - we use `Promise.all` to write those 4 files and in some weird cases it might not be last/4th call (? not sure here if Promise.all guarantees any order of promises, it probably doesn't)

Reply:
It definitely doesn't! If we want a "sequential" promise, there are some ways you can get that with a .reduce() function, but Promise.all very much does _not_ guarantee order!

[Source and further info](https://stackoverflow.com/a/30823708)

Reply:
We don't want that promise to be sequential when actually running `gatsby` (we cold just chain await calls there) - we need reliable way to test if one of function calls was called with expected arguments - but it seems to not work ( https://github.com/gatsbyjs/gatsby/pull/7878#discussion_r215307333 ). Other solution (which I don't like) is to go into `spy.mock.calls` array and inspect that ourselves

Reply:
What about something like this? I think we _have_ to do this because we can't guarantee execution order.

```js
const [,data] = spy.mock.calls.find(([name]) => name === `my/gatsby/project/.cache/data.json.${now}`);

expect(data).toEqual(expectedResult);
```

it's not _great_ but I think it's a little clearer. Also that code may not work exactly _as is_

Reply:
So what do you think about extending `expect` with our own custom matcher?

```js
expect.extend({
  async toHaveBeenCalledOnceWith(received, ...expected) {
    // test that `received` has one of his calls with `expected` arguments
  }
})

// ...

await writePages()

expect(spy).toHaveBeenCalledOnceWith(
  `my/gatsby/project/.cache/data.json.${now}`,
  expectedResult
);
```

Reply:
I was just testing this locally:
```js
it(`tests`, () => {
  const fn = jest.fn()

  fn(`a1`, `b1`)
  fn(`a2`, `b1`)
  fn(`a3`, `b1`)
  fn(`a4`, `b1`)

  expect(fn).toBeCalledWith(`a1`, `b1`)
  expect(fn).toBeCalledWith(`a2`, `b1`)
  expect(fn).toBeCalledWith(`a3`, `b1`)
  expect(fn).toBeCalledWith(`a4`, `b1`)
  expect(fn).not.toBeCalledWith(`a5`, `b1`)
})
```
and this works so not sure why this wouldn't work here as you mentioned previously

Reply:
It seems you're right!! I tried and it work fine. Maybe fails was caused by something else 🤔. Sorry for making you waste your time 😞.

Reply:
I just push a new fixup. I took the opportunity to split the test in two parts.